fineract-provider/src/main/java/org/apache/fineract/infrastructure/security/api/AuthenticationApiResource.java

```diff
@@ -158,8 +167,54 @@ public String authenticate(@Parameter(hidden = true) final String apiRequestBody
                         returnClientList ? clientReadPlatformService.retrieveUserClients(userId) : null);
             }
 
+            String clientIp = "Unknown IP Address";
+            if (servletRequest != null) {
+                clientIp = servletRequest.getHeader("X-Forwarded-For");
+                if (clientIp == null || clientIp.isEmpty()) {
+                    clientIp = servletRequest.getRemoteAddr();
+                } else {
+                    // The X-Forwarded-For header can contain multiple IP addresses, in case of proxies.
+                    // The first IP in the list is the original client.
+                    clientIp = Iterables.get(Splitter.on(',').split(clientIp), 0);
+                    ;
+                }
+            }
+            this.jdbcTemplate.update("insert into m_portfolio_command_source "
+                    + "(action_name,entity_name,office_id,api_get_url,command_as_json,resource_id,maker_id,made_on_date,processing_result_enum) "
+                    + "values(?, ?,?,?,?,?,?,current_date,1) ", "LOGIN", "AUTHENTICATION", principal.getOffice().getId(),
+                    "/authenticate/login", "{ipAddress:\"" + clientIp + "\"}", userId, userId);
         }
 
         return this.apiJsonSerializerService.serialize(authenticatedUserData);
     }
+
+    @Path("/logout")
+    @POST
+    @Consumes({ MediaType.APPLICATION_JSON })
+    @Produces({ MediaType.APPLICATION_JSON })
+    public String processLogout(final String apiRequestBodyAsJson, @Context HttpServletRequest servletRequest) {
+        final CommandWrapper commandRequest = new CommandWrapperBuilder().invalidateTwoFactorAccessToken().withJson(apiRequestBodyAsJson)
+                .build();
+
+        String clientIp = "Unknown IP Address";
+        if (servletRequest != null) {
+            clientIp = servletRequest.getHeader("X-Forwarded-For");
+            if (clientIp == null || clientIp.isEmpty()) {
+                clientIp = servletRequest.getRemoteAddr();
+            } else {
+                // The X-Forwarded-For header can contain multiple IP addresses, in case of proxies.
+                // The first IP in the list is the original client.
+                clientIp = Iterables.get(Splitter.on(',').split(clientIp), 0);
+                ;
+            }
+        }
+
+        AppUser appUser = this.springSecurityPlatformSecurityContext.authenticatedUser();
```

Comment:
This should be at the service layer also . 
perhaps a single method can handle both where LOGIN and LOGOUT are events